README.MD

<h1 align="center">
  <a href="https://github.com/volodiapg/giraff">
    <img src="giraffe.jpg" alt="GIRAFF mascot" width="320" height="320" style="border-radius: 50%;box-shadow: rgba(50, 50, 93, 0.25) 0px 30px 60px -12px, rgba(0, 0, 0, 0.3) 0px 18px 36px -18px;">
  </a>
</h1>

<div align="center">
  <h1>
    GIRAFF: Reverse auction-based placement for fog functions 
  </h1>
  <h3>
  Using FaaS functions and auctions for a sustainable fog model
  </h3>
  <!-- <br />
  <br />
  <a href="">Report a Bug</a>
  ·
  <a href="">Request a Feature</a>
  .
  <a href="">Ask a Question</a> -->
</div>

<div align="center">
<br />

[![license](https://img.shields.io/badge/License-MIT-success?style=flat-square)](LICENSE)

[![made with hearth by VolodiaPG](https://img.shields.io/badge/Made%20with%20%E2%99%A5%20by-VolodiaPG-%23ff0000?style=flat-square)](https://github.com/volodiapg)

</div>

<details open="open">
<summary>Table of Contents</summary>

- [About](#about)
  - [Built With](#built-with)
- [Getting Started](#getting-started)
  - [Overview](#overview)
  - [Prerequisites](#prerequisites)
  - [Software](#software)
  - [VMs](#vms)
  - [Experiments](#experiments)
  - [Processing the results](#processing-the-results)
  - [Notes about the system architectures](#notes-about-the-system-architectures)
- [Contribution or utilization](#contribution-or-utilization)
- [License](#license)
- [Acknowledgements](#acknowledgements)

</details>

## About

Function-as-a-Service (FaaS) applications could harness the disseminated nature of the fog and take advantage of fog’s benefits, such as real-time processing and reduced bandwidth. The FaaS programming paradigm allows applications to be divided in independent units called “functions.” However, deciding how to place those units in the fog is challenging. Fog contains diverse, potentially resource-constrained nodes, geographically spanning from the Cloud to the IP network edges. These nodes must be efficiently shared between the multiple applications that will require to use the fog.

We introduce “fog node ownership,” a concept where fog nodes are owned by different actors that chose to give computing resources in exchange for remuneration. This concept allows for the reach of the fog to be dynamically extended without central supervision from a unique decision taker, as currently considered in the literature. For the final user, the fog appears as a single unified FaaS platform. We use auctions to incentivize fog nodes to join and compete for executing functions.

Our auctions let Fog nodes independently put a price on candidate functions to run. It introduces the need of a “Marketplace,” a trusted third party to manage the auctions. Clients wanting to run functions communicate their requirements using Service Level Agreements (SLA) that provide guarantees over allocated resources or the network latency. Those contracts are propagated from the Marketplace to a node and relayed to neighbors.

Key features of Global Integration of Reverse Auctions and Fog Functions (**GIRAFF**):

- Standard stack (Kubernetes, OpenFaaS)
- Efficient programming in Rust to enable effective collaboration (also cross-compiling in the future?)
- Nix to reproduce scientifically the experiments and maintain the same development environment for everyone
- Functions to deploy
- [Grid’5000](https://www.grid5000.fr/w/Grid5000:Home) support thanks to [EnosLib](https://discovery.gitlabpages.inria.fr/enoslib/)

<details>
<summary>Additional info</summary>
<br>

This project has been started as an internship sponsored as I was a student from National Institute of Applied Sciences Rennes (INSA Rennes) and a second master (SIF) under University of Rennes 1, University of Southern Brittany (UBS), ENS Rennes, INSA Rennes and CentraleSupélec.

A thesis is being financed by the «Centre INRIA de l’Université de Rennes» to pursue the work.

</details>

### Built With

- Nix
- Rust
- Go
- Kubernetes (K3S)
- OpenFaaS
- EnosLib

## Getting Started

This repo uses extensively [`just`](https://github.com/casey/just) as a powerful CLI facilitator.

### Overview

Here is an overview of the content of this repo:

```shell
.
├── testbed # Contains EnosLib code to interact with Grid'5000 (build + deployment of live environment at true scale)
├── manager # contains the code of the marketplace and the fog_node
├── iot_emulation # Sends request to fog nodes in the experiments to measure their response time, etc.
└── openfaas-functions # contains code of Fog FaaS functions
```

### Prerequisites

1. Install [Nix](https://nixos.org/download#download-nix)
    > [Nix portable](https://github.com/DavHau/nix-portable) is not officially supported, however it may be enough to explore the project without installing Nix
2. Append to /etc/nix/nix.conf:
    ```nix
    extra-experimental-features = nix-command flakes
    max-jobs = auto
    cores = 0
    log-lines = 50
    builders-use-substitutes = true
    trusted-useres = root <YOUR USERNAME>
    ```
    > This enables commands such as `nix develop`, multithreading, bigger logs and the usage of the projet's [cachix](https://giraff.cachix.org) cache
3. `nix develop` (or configure direnv)
4. All usual commands can be found in the justfiles, just type `just --list`

> These commands work when `flake.nix` and `justfile` are present in the current directory you are in.

### Software

This section concerns OpenFaaS functions, the code for the fog nodes under `manager/`, the code for `iot_emulation`, the code for the proxy.

Usually, `nix develop` gets you started. I code using VS Code, thus the extension [Direnv](https://marketplace.visualstudio.com/items?itemName=mkhl.direnv) will make VS Code use all the applications/env loaded in the `nix develop`, e.g. you can use Rust/Golang LSPs server/toolchains inside VS Code without ever “installing” them on the computer.

### VMs

VMs detail are located in `testbed/iso/flake.nix`. There lies the whole configuration of the experimental VM. In `testbed/flake.nix` you can also see the VM used to deploy the code in grid’5000. 

To start the same vm as in the experiments, one should go to `testbed/iso` and enter a `nix develop`. Then starting the VM is a matter of `just vm`. Once the VM is started, connection can be made with `just ssh-in`.
> This process is used to start the VM used to develop the fog node software programs

### Experiments

This uses the exact same VMs as the previous section. To generate the VM disk and send it to grid’5000, enter `just upload <rennes|nancy|...>`. The disk will be uploaded.

Then go to `testbed`, once this is done, you can configure the experiments in `.env`; `.experiments.env` handles the variations of multiple runs. These values are used in `integration.py` that handles the deployment and `definitions.py` that defines the fog network and some Kubernetes configurations.

Then, `just upload` will rsync both the master’s VM and the previously described files to the configured grid’5000 cluster.

Finally, `just master_exec <ghcr.io username> <experiment name>` will start the experiment as configured.

> Note that in `.experiments.env` there are some options to gracefully handle failures, as I use GNU parallel, one can “resume” a job that had some failures before.

In the end, experimental results will be available on the cluster in `metric-arks`. One is able to get them back using the command `just get_metrics_back`. This will download them in the local `metrics-arks` folder.

#### Relation to Azure traces

Azure FaaS traces have been released in 2020. Those traces have been characterized by probabilistic laws [[Hernod](https://doi.org/10.1145/3542929.3563468)]. Those laws are described in the following:

- Execution times follow a highly-variable Log-normal law;
- functions live in the range of milliseconds to minutes; 
- ~~Functions are billed with a millisecond granularity;~~
- the median execution time if 600ms;
- the 99%th execution time is more than 140 seconds;
- 0.6% of functions account for 90% of total invocations (they also test a multiple functions balanced workload, where each function receives the same load);
- arrivals follow an open-loop Poisson law;
- arrival burstiness index is -0.26;
- total number of invocations does not vary much;
- invocations follow diurnal patterns as the Cloud does too;
- functions are busy-spun for the execution_duration, repeating a timed math operation over-and-over.

In their simulations/experiments, they state they use:

- mu = -0.38
- sigma = 2.36

For their experiments, they use MSTrace and select a subset of real traces to be re-run on the FaaS platform.
Functions are 256MB of RAM.

#### Distribution for fog specific characteristics

- A Student T distribution is used for function latencies (as we don't know the real sample size nor the std deviation). The distribution would be divided in three buckets: functions that do not care about the latency (meaning the threshold t is at 10 secs), functions that are normal: they would want their responses back in a usable time (t = 150ms), and low latency functions (t = 15ms). The two extrema would be 5% of the total number functions. We use df=10 with -2.25 and 2.25 for the extrema.

### Included functions

Our own functions are:
  
  - `echo` simply register to Influx the arrival time of the functions. It can transfer the request to a next function. (Rust)

We borrowed functions from [EdgeFaaSBench](https://github.com/kaustubhrajput46/EdgeFaaSBench)
  
  - `speech-to-text` (Python)

### Processing the results

This flake has two modes: `just labExport` will start a JupyterLab with Latex support and Tikz export for the graphs, `just lab` starts a lighter version without Latex and Tikz.

Then, data exploitation is done using R inside the JupyterLab server.

> With article submissions, one can find the raw data in the [Release page](https://github.com/volodiapg/faas_fog/releases/latest). Take the latest, extract it to `testbed` under a directory named `metrics-arks`.  Then run `just lab` and you will be able to explore the data. Please notice that this process is heavy on the CPU and especially the RAM. I used some Systemd magic to prevent my computer from using too much RAM and cut the program if so.

### Dev

To locally develop, I will describe the simple steps to get started:
- start the VM: `cd testbed/iso; just vm`
- start the iot_emulation: `cd iot_emulation; just run`
- start the manager (fog node && market): `cd manager; just run <ip:local ip(not localhost)>`
- upload the functions to the registry: `cd openfaas-functions; just`
- you can run parts of the experimental configuration using `cd manager; just expe <ip:same as before>`


### Notes about the system architectures

Most of the flakes are compatible with both Linux and macOS. However, when generating packages for Linux (like the VM), only Linux can. Extension could be done to enable full cross-platform support.

## Contribution or utilization

Please open an issue or contact me from the info in my [GitHub profile](https://github.com/volodiapg) so that I may be of assistance.

## License

This project is licensed under the **MIT license**.

See [LICENSE](LICENSE) for more information.

## Acknowledgements

Thanks for these awesome resources that were used during the development of this project

- [GNU Parallel](https://www.gnu.org/software/parallel/)
- [EnosLib](https://discovery.gitlabpages.inria.fr/enoslib/index.html)
- [Grid’5000](https://www.grid5000.fr/w/Grid5000:Home)


CHANGELOG.md
- - -
## 2.2.0 - 2023-11-17
#### Features
- add function chaining + nixifying and caching - (c55de52) - Volodia PAROL-GUARINO
- add env vars to the sla and function - (d9d1215) - Volodia PAROL-GUARINO
#### Miscellaneous Chores
- migrate from unstable to stable channel - (8b4bd1d) - Volodia PAROL-GUARINO
- update repo name in some links - (25b5c9c) - Volodia PAROL-GUARINO

- - -

## 2.1.0 - 2023-10-30
#### Features
- replace cargo2nix with crane and fix build steps for jupyterlab and rust - (7effeb8) - Volodia PAROL-GUARINO

- - -

## 2.0.0 - 2023-10-11
#### Bug Fixes
- **(testbed)** Reversed nixpkgs to previous working version to make the testbed run again - (94c4c4b) - Volodia PAROL-GUARINO
- make changelog not bugged - (1a5afeb) - Volodia PAROL-GUARINO
- fix update that broke stuff (reverted) - (1a2ec29) - Volodia PAROL-GUARINO
- fix paths not updated after refactor - (2a6c005) - Volodia PAROL-GUARINO
#### Features
- use kubenix for reproducible deployment of openfaas - (dc4f2e9) - Volodia PAROL-GUARINO
#### Miscellaneous Chores
- **(update)** update lock - (4d7403c) - Volodia PAROL-GUARINO
- **(update)** fix poetry2nix having issues with nixpkgs - (1c03742) - Volodia PAROL-GUARINO
- **(update)** update lock - (1f93b89) - Volodia PAROL-GUARINO
- remove already ignored dir - (fe9925b) - Volodia PAROL-GUARINO
- fix missing paths for devshells - (c613c15) - Volodia PAROL-GUARINO
- remove unused vpn options - (1058971) - Volodia PAROL-GUARINO
#### Refactoring
- Unify toolchains - (555065e) - Volodia PAROL-GUARINO
- Unify flakes - (034ac3b) - Volodia PAROL-GUARINO
- move enos code into testbed dir - (baced4a) - Volodia PAROL-GUARINO

- - -

## 1.0.0 - 2023-10-04
#### Bug Fixes
- **(echo-function)** better to sort the buckets before remove duplicates, as it is required per the documentation - (4e54f34) - Volodia PAROL-GUARINO
- **(experiment)** use memsuspend instead of memfree + fix missing SIZE_MULTIPLIER env var - (a0ff410) - Volodia PAROL-GUARINO
- **(experiment)** Fix networking between iot_emulation and proxies - (f7a9851) - Volodia PAROL-GUARINO
- **(experiment)** Move copy of env and definition files before parralelizing them - (fbf66e1) - Volodia PAROL-GUARINO
- **(experiment)** Fix enoslib version to use SMT by default + small fixes - (bbebcc7) - Volodia PAROL-GUARINO
- **(experiment)** Allow for specification of a different local port - (54a7912) - Volodia PAROL-GUARINO
- **(experiment)** Add sensible default env variables to run outside a container - (1825544) - Volodia PAROL-GUARINO
- **(experiment)** Fix envrc - (7fec46d) - Volodia PAROL-GUARINO
- **(fog_node)** Fix pricing ratio not set correctly - (8738421) - Volodia PAROL-GUARINO
- **(fog_node)** Remove function mem+cpu utilization from the metrics instead of adding them, again, when unprovisioning a function - (3a4646f) - Volodia PAROL-GUARINO
- **(fog_node)** Add function tracking repository in the dynamic dispatch of actix to be able to actually measure something - (6304218) - Volodia PAROL-GUARINO
- **(functions)** Fix unmatching datatypes in dynamic dispatch - (b625bed) - Volodia PAROL-GUARINO
- **(functions)** Go back to only using an histogram - (97ea26f) - Volodia PAROL-GUARINO
- **(functions)** Fix histograms throwning errors because vector was not an ordered set - (fe478f2) - Volodia PAROL-GUARINO
- **(functions)** Fix micondifurations in flake - (3c648be) - Volodia PAROL-GUARINO
- **(iot_emulation)** Missing loading of nix-generated container image that was not updated on any remote because of this - (a29ace4) - Volodia PAROL-GUARINO
- **(manager)** Fix DOCKER env variable not being settable by the user - (4a8b8ec) - Volodia PAROL-GUARINO
- **(manager)** Fix nix develshell - (aeb83b3) - Volodia PAROL-GUARINO
- **(manager)** Add removal of used resources when functions gets removed - (eb152c1) - Volodia PAROL-GUARINO
- **(manager)** Fix generation for docker images - (92bb9ec) - Volodia PAROL-GUARINO
- **(pre-commit)** Fix wrong path passed to script - (a83caf8) - Volodia PAROL-GUARINO
- **(simulator)** fix missing level not being displayed - (65c6ed5) - Volodia PAROL-GUARINO
- **(simulator)** fix incorrect use of sorted that doesn't sort in place - (f61267b) - Volodia PAROL-GUARINO
- **(simulator)** fix price monitoring not working - (fa64bc6) - Volodia PAROL-GUARINO
- remove junk files - (e4ab548) - Volodia PAROL-GUARINO
- Add csv and bak files to the gitignore - (2bcd97c) - Volodia PAROL-GUARINO
#### Documentation
- **(experiment)** add legends for graphs - (32d5fd2) - Volodia PAROL-GUARINO
- add link to released raw data - (cce7d7d) - Volodia PAROL-GUARINO
- Change h2 to h3 - (facc996) - Volodia PAROL-GUARINO
- add mascot - (e4a0e88) - Volodia PAROL-GUARINO
- update ToC - (3729d1a) - Volodia PAROL-GUARINO
- update readme - (6c67401) - Volodia PAROL-GUARINO
#### Features
- **(README)** Update - (d392397) - Volodia PAROL-GUARINO
- **(echo-function)** Add debug entry to test locally - (8a83aba) - Volodia PAROL-GUARINO
- **(experience)** Add .env file support + load settings support - (03880e0) - Volodia PAROL-GUARINO
- **(experiment)** Add working experiment deployment - (57ed654) - Volodia PAROL-GUARINO
- **(experiment)** run experiments from inside g5k to fix the shitstorm running outside actually is... - (c5bae44) - Volodia PAROL-GUARINO
- **(experiment)** add VM generation for running experiments from g5k - (ef7c4af) - Volodia PAROL-GUARINO
- **(experiment)** add dry experiemnt to measure the number of used nodes without deploying - (f90d7e0) - Volodia PAROL-GUARINO
- **(experiment)** add support for randomly generated networks in realworld deployments - (05ca502) - Volodia PAROL-GUARINO
- **(experiment)** add retries and timeout for experiments - (b6c583e) - Volodia PAROL-GUARINO
- **(experiment)** Add support for parallel and configurable experiments - (e561bf9) - Volodia PAROL-GUARINO
- **(experiment)** add randomizer for size of virtual machines - (4df6b39) - Volodia PAROL-GUARINO

testbed/expe.py
import asyncio
import json
import math
import os
import random
from dataclasses import dataclass, field
from typing import Any, List, Optional

import aiohttp  # type: ignore
import dill
import marshmallow_dataclass  # type: ignore
import numpy as np  # type: ignore
from alive_progress import alive_bar  # type: ignore

port_int = int
MiB_int = int
millicpu_int = int
ms_int = int
secs_int = int


@dataclass
class Function:
    target_node: str
    mem: MiB_int
    cpu: millicpu_int
    latency: ms_int
    cold_start_overhead: ms_int
    stop_overhead: ms_int
    duration: ms_int
    docker_fn_name: str
    function_name: str
    first_node_ip: str | None
    request_interval: ms_int
    arrival: secs_int
    req_content: str


@dataclass
class FunctionPipeline:
    image: str
    nextFunction: Optional[str] = None
    mem: MiB_int = 256
    cpu: millicpu_int = 100


@dataclass
class FunctionPipelineDescription:
    name: str
    content: str
    nbVarName: str
    pipeline: dict[str, FunctionPipeline]


@dataclass
class FunctionProvisioned:
    faas_ip: str
    faas_port: int
    function_id: str
    node_id: str


RANDOM_SEED = os.getenv("RANDOM_SEED")
if RANDOM_SEED is not None and RANDOM_SEED != "":
    random.seed(int(RANDOM_SEED))


def generate_rand(min: int, max: int) -> int:
    # Generate a random number of nbytes
    if RANDOM_SEED is None:
        random_number = os.urandom(4)

        # Convert the bytes to an integer
        random_integer = int.from_bytes(random_number, byteorder="big")

        return (random_integer % (max - min + 1)) + min

    return random.randint(min, max)


def open_loop_poisson_process(rate, time):
    """
    Simulates an open-loop Poisson process.

    Parameters:
    rate: The average rate of events (lambda).
    time: The total time to simulate.

    Returns:
    A list of event times.
    """
    num_events = np.random.poisson(rate * time)
    event_times = np.random.uniform(0, time, num_events)

    # # Sort the event times
    # event_times.sort()

    return event_times


if __name__ == "__main__":
    for k, v in os.environ.items():
        if k in [
            "TARGET_NODES",
            "TARGET_NODE_NAMES",
            "IOT_IP",
            "MARKET_IP",
            "MARKET_LOCAL_PORT",
            "IOT_LOCAL_PORT",
            "EXPE_SAVE_FILE",
            "EXPE_LOAD_FILE",
        ]:
            print(f"{k}={v}")

TARGET_NODES = os.getenv("TARGET_NODES", "").split()
TARGET_NODE_NAMES = os.getenv(
    "TARGET_NODE_NAMES", ""
).split()  # Shoud be in the same order than TARGET_NODES
if len(TARGET_NODES) != 0:
    assert len(TARGET_NODES) == len(TARGET_NODE_NAMES)

IOT_IP = os.getenv("IOT_IP")
MARKET_IP = os.getenv("MARKET_IP")
MARKET_LOCAL_PORT = port_int(os.environ["MARKET_LOCAL_PORT"])
IOT_LOCAL_PORT = port_int(os.environ["IOT_LOCAL_PORT"])
NODES_IP = os.getenv("NODES_IP")
IMAGE_REGISTRY = os.environ["IMAGE_REGISTRY"]

NO_LATENCY = ms_int(os.environ["NO_LATENCY"])
HIGH_LATENCY = ms_int(os.environ["HIGH_LATENCY"])
LOW_LATENCY = ms_int(os.environ["LOW_LATENCY"])

FUNCTION_COLD_START_OVERHEAD = ms_int(os.environ["FUNCTION_COLD_START_OVERHEAD"])
FUNCTION_STOP_OVERHEAD = ms_int(os.environ["FUNCTION_STOP_OVERHEAD"])
EXPERIMENT_DURATION = secs_int(os.environ["EXPERIMENT_DURATION"])

# Debug function for running local
OVERRIDE_FUNCTION_IP = os.getenv("OVERRIDE_FUNCTION_IP")
print(f"OVERRIDE_FUNCTION_IP={OVERRIDE_FUNCTION_IP}")
OVERRIDE_FIRST_NODE_IP = os.getenv("OVERRIDE_FIRST_NODE_IP")
print(f"OVERRIDE_FIRST_NODE_IP={OVERRIDE_FIRST_NODE_IP}")

FUNCTION_DESCRIPTIONS = os.getenv(
    "FUNCTION_DESCRIPTIONS", ""
).split()  # Shoud be in the same order than TARGET_NODES
assert len(FUNCTION_DESCRIPTIONS) != 0


class AsyncSession:
    def __init__(self):
        self.timeout = aiohttp.ClientTimeout()

    async def __aenter__(self):
        self.session = aiohttp.ClientSession(timeout=self.timeout)
        return self.session

    async def __aexit__(self, exc_type, exc_value, traceback):
        await self.session.close()


async def put_request_fog_node(function: Function):
    url = f"http://{MARKET_IP}:{MARKET_LOCAL_PORT}/api/function"
    headers = {"Content-Type": "application/json"}
    duration = function.cold_start_overhead + function.duration + function.stop_overhead
    data = {
        "sla": {
            "memory": f"{function.mem} MB",
            "cpu": f"{function.cpu} millicpu",
            "latencyMax": f"{function.latency} ms",
            "maxReplica": 1,
            "duration": f"{duration} ms",
            "functionImage": f"{IMAGE_REGISTRY}/{function.docker_fn_name}",
            "functionLiveName": f"{function.function_name}",
            "dataFlow": [  # TODO: update because outdated
                {
                    "from": {"dataSource": f"{function.target_node}"},
                    "to": "thisFunction",
                }
            ],
        },
        "targetNode": f"{function.target_node}",
    }
    async with AsyncSession() as session:
        async with session.put(url, headers=headers, json=data) as response:
            http_code = response.status
            response = await response.content.read()
            return response, http_code


async def post_request_chain_functions(urls: List[FunctionProvisioned]):
    headers = {"Content-Type": "application/json"}
    last = len(urls) - 1
    if last == 0:
        return None

    await asyncio.sleep((FUNCTION_COLD_START_OVERHEAD / 2) / 1000)

    print(urls)

    ret = []
    for ii in range(0, last):
        faas_ip = urls[ii].faas_ip
        if OVERRIDE_FUNCTION_IP is not None:
            faas_ip = OVERRIDE_FUNCTION_IP
        data = {
            "nextFunctionUrl": f"http://{faas_ip}:{urls[ii+1].faas_port}/function/fogfn-{urls[ii+1].function_id}",
        }
        # Sync request
        url = f"http://{urls[ii].faas_ip}:{urls[ii].faas_port}/function/fogfn-{urls[ii].function_id}/reconfigure"

        print(
            "from:",
            urls[ii].function_id,
            "- to:",
            urls[ii + 1].function_id,
            "- url:",
            url,
            "- data:",
            data,
        )
        async with AsyncSession() as session:
            async with session.post(url, headers=headers, json=data) as response:
                http_code = response.status
                response = await response.content.read()

                ret.append((response, http_code))

    return ret


async def put_request_iot_emulation(
    provisioned: FunctionProvisioned,
    function: Function,
):
    faas_ip = function.first_node_ip
    if OVERRIDE_FIRST_NODE_IP is not None:
        faas_ip = OVERRIDE_FIRST_NODE_IP

    url = f"http://{IOT_IP}:{IOT_LOCAL_PORT}/api/cron"
    headers = {"Content-Type": "application/json"}
    data = {
        "iotUrl": f"http://{IOT_IP}:{IOT_LOCAL_PORT}/api/print",
        "nodeUrl": f"http://{provisioned.faas_ip}:{provisioned.faas_port}/function/fogfn-{provisioned.function_id}",
        "functionId": provisioned.function_id,
        "tag": function.function_name,
        "initialWaitMs": function.cold_start_overhead,
        "durationMs": function.duration,
        "intervalMs": function.request_interval,
        "firstNodeIp": faas_ip,
        "content": function.req_content,
    }

    async with AsyncSession() as session:
        async with session.put(url, headers=headers, json=data) as response:
            http_code = response.status
            response = await response.content.read()
            return response, http_code


async def register_new_functions(functions: List[Function]) -> bool:
    await asyncio.sleep(functions[0].arrival)

    responses = []
    for ii in range(0, len(functions)):
        function = functions[ii]
        response, code = await asyncio.ensure_future(put_request_fog_node(function))
        if code != 200:
            print("Fog request somehow failed", code, response)
            return False

        response = json.loads(response)
        faas_ip = response["chosen"]["ip"]
        node_id = response["chosen"]["bid"]["nodeId"]
        faas_port = response["chosen"]["port"]
        function_id = response["chosen"]["bid"]["id"]
        response = FunctionProvisioned(faas_ip, faas_port, function_id, node_id)
        responses.append(response)
        if ii + 1 < len(functions):
            functions[ii + 1].target_node = response.node_id
            functions[ii + 1].first_node_ip = response.faas_ip

        print(f"Provisioning... {ii+1}/{len(functions)}")

    print(f"Provisioned {','.join([ff.function_name for ff in functions])}")

    tasks = []
    tasks.append(asyncio.create_task(post_request_chain_functions(responses)))
    tasks.append(
        asyncio.create_task(
            put_request_iot_emulation(
                responses[0],
                function=functions[0],
            )
        )
    )

    response_list = await asyncio.gather(*tasks)
    responses_chain = response_list[-2]
    response_iot, code_iot = response_list[-1]

    if responses_chain is not None:
        for response, http_code in responses_chain:
            if http_code != 200:
                print("Request failed", http_code, response)
                return False

    print(f"Chained {','.join([ff.function_name for ff in functions])}")
    if code_iot == 200:
        print(f"Registered cron for {functions[0].function_name}")
        return True

    print("---\n", response_iot.decode("utf-8").replace("\\n", "\n"))

    return False


successes = 0
errors = 0


async def do_request_progress(bar: Any, functions: List[Function]):
    global successes
    global errors
    success = await register_new_functions(functions)
    if success:
        successes += 1
    else:
        errors += 1
    bar.text = f"--> Done {successes}, failed to provision {errors} functions..."
    bar()


PERCENTILE_NORMAL_LOW = -2
PERCENTILE_NORMAL_HIGH = 2


def load_function_descriptions() -> List[FunctionPipelineDescription]:
    ret = []
    schema = marshmallow_dataclass.class_schema(FunctionPipelineDescription)()
    for desc_file in FUNCTION_DESCRIPTIONS:
        with open(desc_file) as ff:
            ret.append(schema.load(json.load(ff)))
    return ret


async def save_file(filename: str):
    functions: List[List[Function]] = []

    function_descriptions = load_function_descriptions()
    nb_functions = []
    for fn_desc in function_descriptions:
        nb_functions.append(int(os.getenv(fn_desc.nbVarName, "0")))
    for target_node_name in TARGET_NODE_NAMES:
        for ii, fn_desc in enumerate(function_descriptions):
            nb_function = nb_functions[ii]
            latencies = [
                max(1, math.ceil(x)) for x in np.random.normal(70, 30.0, nb_function)
            ]
            request_intervals = [
                math.ceil(abs(1000 * x))
                for x in np.random.lognormal(-0.38, 2.36, nb_function)
            ]
            # TODO check that thing
            durations = [
                math.ceil(1000000 * x)
                for x in np.random.lognormal(-0.38, 2.36, nb_function)
            ]
            arrivals = [
                math.ceil(x)
                for x in open_loop_poisson_process(nb_function, EXPERIMENT_DURATION)
            ]

            for index in range(0, nb_function):
                latency = latencies[index]
                arrival = arrivals[index]
                duration = durations[index]
                request_interval = request_intervals[index]

                fn_name = list(fn_desc.pipeline.keys())[0]

                print(fn_name)
                print(fn_desc)

                fn_chain = list()
                while True:
                    fn: FunctionPipeline = fn_desc.pipeline[fn_name]
                    function_name = (
                        f"{fn_name}"
                        f"-i{index}"
                        f"-c{fn.cpu}"
                        f"-m{fn.mem}"
                        f"-l{latency}"
                        f"-a{arrival}"
                        f"-r{request_interval}"
                        f"-d{duration}"
                    )
                    fn_chain.append(
                        Function(
                            target_node=target_node_name,
                            mem=fn.mem,
                            cpu=fn.cpu,
                            latency=latency,
                            cold_start_overhead=FUNCTION_COLD_START_OVERHEAD,
                            stop_overhead=FUNCTION_STOP_OVERHEAD,
                            duration=duration,
                            request_interval=request_interval,
                            function_name=function_name,
                            docker_fn_name=fn.image,
                            first_node_ip=None,
                            arrival=arrival,
                            req_content=fn_desc.content,
                        )
                    )

                    if fn.nextFunction == None:
                        break
                    fn_name = fn.nextFunction
                functions.append(fn_chain)

    print(functions)

    with open(filename, "wb") as outp:  # Overwrites any existing file.
        dill.dump(functions, outp, dill.HIGHEST_PROTOCOL)


def load_functions(filename) -> List[List[Function]]:
    functions = []
    with open(filename, "rb") as inp:
        functions = dill.load(inp)
    return functions


async def load_file(filename: str):
    nodes = {}
    for ii in range(len(TARGET_NODE_NAMES)):
        nodes[TARGET_NODE_NAMES[ii].replace("'", "")] = TARGET_NODES[ii].replace(
            "'", ""
        )
    functions = load_functions(filename)
    tasks = []
    bar_len = len(functions)

    response = None
    async with AsyncSession() as session:
        async with session.get(
            f"http://{MARKET_IP}:{MARKET_LOCAL_PORT}/api/fog"
        ) as response:
            response.status
            response = await response.json()
    fognet = {fog_node_data["id"]: fog_node_data["ip"] for fog_node_data in response}

    with alive_bar(bar_len, title="Functions", ctrl_c=False, dual_line=True) as bar:
        for function_sublist in functions:
            # Use the id of the node instead of its name
            for function in function_sublist:
                function.target_node = nodes[function.target_node.replace("'", "")]
                function.first_node_ip = (
                    NODES_IP if NODES_IP else fognet[function.target_node]
                )
            tasks.append(
                asyncio.create_task(
                    do_request_progress(bar=bar, functions=function_sublist)
                )
            )

        await asyncio.gather(*tasks)


async def main():
    if os.getenv("EXPE_SAVE_FILE") is not None:
        await save_file(os.getenv("EXPE_SAVE_FILE"))
    elif os.getenv("EXPE_LOAD_FILE") is not None:
        print(f"Using market ({MARKET_IP}) and iot_emulation({IOT_IP})")
        await load_file(os.getenv("EXPE_LOAD_FILE"))
        print(f"--> Did {successes}, failed to provision {errors} functions.")

    else:
        print("Not EXPE_SAVE_FILE nor EXPE_LOAD_FILE were passed, aborting")
        return 1


if __name__ == "__main__":
    asyncio.run(main())
